### DISCUSSION TITLE
MS2130 grabber. Summary and firmware testing

### DISCUSSION THREAD

**Discussion by awawa-dev** [awawa-dev/HyperHDR#729] [Q&A]
This topic is a kind of summary of information from the previous topic (https://github.com/awawa-dev/HyperHDR/discussions/499), which has grown a lot over the past year and it is difficult to find what is most important.

> [!WARNING]
> The most serious problems with ms2130:
> - different firmwares with different image quality
> - frequent sudden decrease in brightness in the captured stream e.g. when switching HDMI sources
> - problem with YUV limited used by e.g. NVidia Shield. Affects both SDR and HDR
>
> It seems that we can remedy this and standardize it by uploading our experimental firmware (with a modified EDID) and at the same time solve the above problems of variable brightness and limited YUV mode.

1. I have & test probably the most popular versions of MS2130 grabbers. Model A is closed and will be rather difficult to open, if at all possible. Model B is barrel-shaped and resembles the older ms2109. Both models I received had the same firmware.  \
<img src="https://user-images.githubusercontent.com/69086569/218528421-8225caba-faaf-4e10-badd-a5e5b184be2a.jpg" width=240px/><img src="https://user-images.githubusercontent.com/69086569/217898940-1532cadb-a974-4924-b292-a61ceb5025a3.png" width=240px/>

2. SDR and HDR quality is excellent, the latency is very low ~50ms depending on the capture settings. \
\
If we use eg. Ezcoo EZ-SP12H2 splitter, it is also possible to capture the Dolby Vision signal, or at least some versions of it. Contrary to the old MS2109, you don't need to modify contrast, brightness or saturation. Just leave it on defaults which means 0 for contrast/brightness/saturation/hue in HyperHDR grabber configuration. If there were other values set before, you need to restart Raspberry Pi/PC afterwards to apply new settings.

3) ## What you should definitely pay attention to?<br/>
    Unfortunately, subsequent versions of the firmware have introduced numerous changes, and not necessarily positive ones.  \
    **You cannot use HyperHDR LUTs if you have newer firmware. You need to calibrate your grabber yourself.**  \
\
You can recognize the old firmware by the modes offered: when connected to USB2.0 → only MJPEG (and NV12 under Windows), USB3.0 → only YUV. It targets FCC coefs. New firmware offers some YUV modes for USB2.0 and MJPEG modes for USB3.0 but at a great price that is definitely not worth it. Also HyperHDR calibration log result will detect Rec.709 for new firmware. It is nothing wrong in Rec.709, as other USB grabbers also target it, but it was changed without notice and at the same time, as a result of other changes, the image quality including contrast and color gradient decreased. \
\
Since the new MS2130 firmware produces a much brighter image, the LED may not turn off in the event of a black image on the TV screen, even if you have disabled the "Minimum Backlight" option in HyperHDR in the processing tab.

4) If you have an older firmware version, you can install the LUT directly from HyperHDR v20 \
\
<br/><img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/6bc1e748-51c6-4674-bcb9-183481407fd1" width=480px/><br/>
5. ## Is it possible to downgrade the firmware or upload another experimental version?
    In the case of the A and B models I have, this was possible. However, flashing the firmware is always associated with risk, and there is no guarantee that in subsequent revisions of these grabbers, manufacturers will introduce breaking changes that will cause firmware incompatibility: you do it at your own risk and we do not provide any warranty. __Read more here__: [link](https://github.com/awawa-dev/HyperHDR/discussions/729#discussioncomment-8126044)

6. ## When using it with the Ezcoo splitter probably due to EDID causing the handshake problem, differences in brightness of the captured non-SDR image were observed.  Solution: 
   - make sure you turn on the devices using always the same sequence (tv/rpi/splitter/grabber the order is only for example)
   - disconnect and reconnect Ezcoo's USB power supply
   - if you want you can try the experimental firmware which seems to solve this problem (contact me @ email)
  
7. ## What about limited YUV modes used e.g. by NVidia Shield?
    You may experience high brightness levels in captured very dark scenes and reduced maximum brightness in captured very bright scenes. Unfortunately the only known way to fix it is to update the ms2130 firmware.

**Comment by awawa-dev** [comment]
# MS2109 & MS2130 Dolby Vision testing. MS2130 brightness issue detected when using Ezcoo splitter.

I finally managed to configure Amazon Firestick 4k so that Nova player can open files via SMB on the network to test Dolby Vision. Tested on promotional materials and trailers provided by @astarom (thanks!). I made sure that my LG TV actually recognizes that the signal is Dolby Vision:

![obraz](https://github.com/awawa-dev/HyperHDR/assets/69086569/c55f6f7c-95a5-4cfe-8a7c-af648e508ff1)

Maybe I'll start from the end because it's the most important:
After completing the ms2109 tests and uploading the new LUT, I connected the ms2130 (model A, old firmware) and the image suddenly became much darker. HDR tone mapping was turned off, but it didn't matter because even with tone mapping enabled the image could be too dark. I stopped the USB grabber in the remote tab, connected the MS2130 B model (experimental firmware with custom EDID based on MS2109), turned on the USB grabber in the remote and the image suddenly became brighter again. I repeated this several times, each time the ms2130A captured a much darker image, as several users had previously described. Switched back whole HyperHDR setup to MS2109 and the image was also fine, just like MS2130B with custom firmware. 

Switched back to MS2130A: the image is again too dark, __power off&on Ezcoo USB power supply and suddenly the image is fine even for MS2130A__.  The difference in brightness captured in the screenshot:
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/5c0621f9-c642-4904-8531-34de41e68575" width=320px/> <img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/d900028c-0746-4b84-b5c8-115cb312120d" width=320px/>
<details>

![shot070145](https://github.com/awawa-dev/HyperHDR/assets/69086569/5c0621f9-c642-4904-8531-34de41e68575|width=100px) ![shot070226](https://github.com/awawa-dev/HyperHDR/assets/69086569/d900028c-0746-4b84-b5c8-115cb312120d|width=100px)

</details>

Below are the test results for Dolby Vision promotional materials using the MS2109 and MS2130, each time with tone mapping turned off and on. Amazon FireStick 4k is connected to Ezcoo EZ-SP12H2 splitter, TV LG C9 is connected to Ezcoo HDMI1, USB grabber is connected to Ezcoo HDMI2, Ezcoo scaller is disabled for HDMI1 and enabled for HDMI2, Ezcoo EDID copy mode is selected.

MS2109:

<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/a9ac0c3a-6db6-4244-867e-6b539bb6e538" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/36266c0c-0f21-45f2-b637-4621b0bd8459" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/9317da7b-d67d-4f85-b212-d5764d6ac131" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/bdf02f69-bda1-4800-bb51-d474d59936ff" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/d8406f6a-b0de-4555-813e-3ff447f12591" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/421f4229-d483-4af5-8064-bf74bb3fae1f" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/e342c6ea-08e2-4a4a-8fab-928e1de72a95" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/4349f2af-8b5f-4d8a-bd61-b9d63219c3e0" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/9ddf2f5c-88e5-4a40-98da-2b5317240b05" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/5546c87f-100e-41c5-8f30-a695bf478f6e" width=320px/>

<details>

![shot063838](https://github.com/awawa-dev/HyperHDR/assets/69086569/a9ac0c3a-6db6-4244-867e-6b539bb6e538)
![shot063833](https://github.com/awawa-dev/HyperHDR/assets/69086569/36266c0c-0f21-45f2-b637-4621b0bd8459)
![shot064243](https://github.com/awawa-dev/HyperHDR/assets/69086569/9317da7b-d67d-4f85-b212-d5764d6ac131)
![shot064247](https://github.com/awawa-dev/HyperHDR/assets/69086569/bdf02f69-bda1-4800-bb51-d474d59936ff)
![shot142221](https://github.com/awawa-dev/HyperHDR/assets/69086569/d8406f6a-b0de-4555-813e-3ff447f12591)
![shot142401](https://github.com/awawa-dev/HyperHDR/assets/69086569/421f4229-d483-4af5-8064-bf74bb3fae1f)
![shot142944](https://github.com/awawa-dev/HyperHDR/assets/69086569/e342c6ea-08e2-4a4a-8fab-928e1de72a95)
![shot142959](https://github.com/awawa-dev/HyperHDR/assets/69086569/4349f2af-8b5f-4d8a-bd61-b9d63219c3e0)
![shot143156](https://github.com/awawa-dev/HyperHDR/assets/69086569/9ddf2f5c-88e5-4a40-98da-2b5317240b05)
![shot143201](https://github.com/awawa-dev/HyperHDR/assets/69086569/5546c87f-100e-41c5-8f30-a695bf478f6e)

</details>

MS2130B

<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/9fa84b06-2c8b-462d-b45e-fedf8403d35f" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/d79889de-74a0-4204-8e7c-3cef3f24bc99" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/998a62dd-c59a-4adb-ad68-cc97847c80ef" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/91a79056-7449-4757-a6f5-1ab7a689d9a8" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/4696b41c-8cdf-49d1-becc-5b993abd1e25" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/57d008b3-2bfb-4112-9f8c-a5a8f37d1b0c" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/d97c163d-8630-4840-84d2-5c0358f0f12e" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/b508da7a-8257-4e71-9adc-ef04f70b6a60" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/8413c0d2-4232-43e7-8486-335889273427" width=320px/>
<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/2f683bba-b44e-4766-b5dc-c48551a2451a" width=320px/>

<details>

![shot071406](https://github.com/awawa-dev/HyperHDR/assets/69086569/9fa84b06-2c8b-462d-b45e-fedf8403d35f)
![shot071410](https://github.com/awawa-dev/HyperHDR/assets/69086569/d79889de-74a0-4204-8e7c-3cef3f24bc99)
![shot071753](https://github.com/awawa-dev/HyperHDR/assets/69086569/91a79056-7449-4757-a6f5-1ab7a689d9a8)
![shot071757](https://github.com/awawa-dev/HyperHDR/assets/69086569/998a62dd-c59a-4adb-ad68-cc97847c80ef)
![shot072015](https://github.com/awawa-dev/HyperHDR/assets/69086569/4696b41c-8cdf-49d1-becc-5b993abd1e25)
![shot072020](https://github.com/awawa-dev/HyperHDR/assets/69086569/57d008b3-2bfb-4112-9f8c-a5a8f37d1b0c)
![shot072252](https://github.com/awawa-dev/HyperHDR/assets/69086569/b508da7a-8257-4e71-9adc-ef04f70b6a60)
![shot072256](https://github.com/awawa-dev/HyperHDR/assets/69086569/d97c163d-8630-4840-84d2-5c0358f0f12e)
![shot072508](https://github.com/awawa-dev/HyperHDR/assets/69086569/8413c0d2-4232-43e7-8486-335889273427)
![shot072513](https://github.com/awawa-dev/HyperHDR/assets/69086569/2f683bba-b44e-4766-b5dc-c48551a2451a)

</details>

## Not every HDMI splitter will provide correct Dolby Vision capture.
If you count on DV support with MS2130, you should probably avoid Feintach, because this is the image I received in the same setup after replacing Ezcoo EZ-SP12H2 with Feintech VSP01201 (also EDID copy).

<img src="https://github.com/awawa-dev/HyperHDR/assets/69086569/2a4ddb0d-bf2c-49c9-8437-43bb74b612e7" width=320px/>

<details>

![shot105123](https://github.com/awawa-dev/HyperHDR/assets/69086569/2a4ddb0d-bf2c-49c9-8437-43bb74b612e7)

</details>

**Reply by amigthea** [reply]
- Dolby Vision content on Amazon FireStick 4k max
- Firestick to HDMI-in (Ezcoo EZ-SP12H2)
- TV LG connected to HDMI-out 1 (Ezcoo EZ-SP12H2)
- MS2130 connected to HDMI-out 2 (Ezcoo EZ-SP12H2)
- The Ezcoo scaler is disabled for HDMI-out 1
- The Ezcoo scaler is enabled for HDMI-out 2
- Ezcoo EDID copy mode is selected.

this is the setup I'm gonna test later today as soon as the splitter arrives. I had the Feintech VSP01201 and can confirm the purpleish tint overlay so I returned it. Why did you replace the ezcoo?

Also, ezcoo amazon page state that:

`When output 4K HDR Dolby Vision pls TURN OFF scaler. NOT down-scale HDR Dolby Vision`

but in your test it worked out right?

**Reply by awawa-dev** [reply]
The Dolby Vision LLDV protocol does not appear to be subject to such limitations as it is very similar to HDR. I have both Ezcoo and Feintech, but Feintech cant enforce LLDV on the source (and no grabber can capture other DV protocols correctly => pink image or no image).

**Comment by awawa-dev** [comment]
# Comparing old & new firmware results (SDR)

In order not to be biased, we will test what the result looks like for each firmware when we capture an SDR image not using HyperHDR but the standard Windows Camera application.

For each color, a section from the left: the oldest firmware, 4k2 firmware (from the article on zhihu.com), the latest MS2130 firmware.
Judge for yourself, but the last two firmware are very similar or identical, much brighter, and something strange/wrong happened to the captured text.

<img src='https://github.com/awawa-dev/HyperHDR/assets/69086569/7001eb61-4a61-4455-870f-389ffb8595b1' width=200px />
<img src='https://github.com/awawa-dev/HyperHDR/assets/69086569/1fe1558b-7294-4f2f-b660-0cb192a7102e' width=200px />
<img src='https://github.com/awawa-dev/HyperHDR/assets/69086569/e2f11018-2918-4c14-b557-0f2cac195a89' width=200px />
<img src='https://github.com/awawa-dev/HyperHDR/assets/69086569/03726305-d961-4c3e-9c51-cb8c93ad3cd4' width=200px />

**Comment by awawa-dev** [comment]
# How to flash MS2130 firmware

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

## Firmware

Unofficial experimental firmware for ms2130 grabber. Use it on your own risk. Provided without warranty of any kind.
Extract it before use.

Version F3 (2024/1/16)
Changelog:
- fixed issue with very dark captured video
- fixed issue when video source uses limited YUV color space
- add missing modes e.g. 1080p24, 1080p120... reported to the display
- ms2130 now reports HDR support
- ms2130 now reports LLDV support

Download: [link to F3 ms2130 firmware](https://disk.yandex.ru/d/AT7MrgJdVN6D7A)
CRC32: 61DA8DD7

## Windows

Windows flasher can be found in many places. Remember that you must first check it with an antivirus program or use Linux and the opensource ms-tools flasher instead.

https://github.com/awawa-dev/HyperHDR/discussions/499#discussioncomment-7811045 
https://github.com/BertoldVdb/ms-tools/issues/11
https://twitter.com/aaaarrrriaaannn/status/1679125963487969282

Example CRC32 (does not mean it is correct!): D919F9DB

Remember to make a backup first.
Manual (Chinese): https://zhuanlan.zhihu.com/p/634052876

## Linux

Requires usage of https://github.com/BertoldVdb/ms-tools
First you must install some packages:
`sudo apt install git libhidapi-dev libudev-dev`

### Use our prebuild _cli_ binary or build it yourself

Download prebuilt _cli_ and extract it to /tmp/ms2130: [cli_aarch64.zip](https://github.com/awawa-dev/HyperHDR/files/14044322/cli_aarch64.zip)

Or build _cli_ yourself (recommended). You need `go` package installed:

https://go.dev/doc/install

For RPi AARCH64 the package is named: go1.21.6.linux-arm64.tar.gz and I assume this in the example below. If you are using different platform or version, it needs to be adjusted. The download link can be found on the mentioned go.dev website.

#### Install `go` first:

<details>

```
cd /tmp
wget https://go.dev/dl/go1.21.6.linux-arm64.tar.gz
sudo rm -rf /usr/local/go && sudo tar -C /usr/local -xzf ./go1.21.6.linux-arm64.tar.gz
export PATH=$PATH:/usr/local/go/bin
go version
```

</details>

#### Build <i>cli</i>

<details>

```
cd /tmp && mkdir ms2130 && cd ms2130
git clone https://github.com/BertoldVdb/ms-tools.git tools
cd tools 
go build -o /tmp/ms2130 ./cli
cd ..
./cli
```

</details>

### Backup & upload new firmware

Make backup of your ms2130 firmware (wait few minutes till it finishes):
`sudo ./cli --log-level=7 read FLASH 0 --filename=/tmp/ms2130/my_backup.bin`

Copy the backup (/tmp/ms2130/my_backup.bin) into safe place.
Put new firmware into /tmp/ms2130 directory for example: /tmp/ms2130/backup.bin

Write new ms2130 firmware (wait few minutes till it finishes):
`sudo ./cli --log-level=7 write-file --verify FLASH 0 /tmp/ms2130/backup.bin`
